### PR TITLE
Avoid exploding memory on transpose error

### PR DESCRIPTION
**Pull request by bcipolli**
Fix for #458.  Now, a bit more checking occurs before allocating memory:
* If a transposed matrix (n_features, n_slices) is passed in, it is transposed before assignment.
* If a 2D matrix with no dimensions of size `n_features` exists, a friendly error is thrown.

### CODE REVIEW

File: nilearn/masking.py [hunk 1]
```diff
@@ -665,7 +665,15 @@ def _unmask_nd(X, mask, order="C"):
...
-        raise ValueError("X must be a 2-dimensional array")
+        raise TypeError("X must be a 2-dimensional array")
+    n_features = mask.sum()
+    if X.shape[-1] != n_features:
+        # Handle (potential) transpose as a special case, as this
+        #   mistake can lead to system-fatal memory allocation below.
+        if X.shape[0] == n_features:
+            X = X.T
+        else:
+            raise TypeError('X must be of shape (samples, %d).' % n_features)
```
Comment (lesteve) [inline nilearn/masking.py]: I'd be enclined to always throw if `X.shape[-1] != n_features` rather than have a special edge case that automagically works.

File: nilearn/masking.py [hunk 2]
```diff
@@ -665,7 +665,15 @@ def _unmask_nd(X, mask, order="C"):
     if mask.dtype != np.bool:
         raise ValueError("mask must be a boolean array")
     if X.ndim != 2:
-        raise ValueError("X must be a 2-dimensional array")
+        raise TypeError("X must be a 2-dimensional array")
+    n_features = mask.sum()
+    if X.shape[-1] != n_features:
+        # Handle (potential) transpose as a special case, as this
+        #   mistake can lead to system-fatal memory allocation below.
+        if X.shape[0] == n_features:
```
Comment (AlexandreAbraham) [inline nilearn/masking.py]: I don't like this kind of magic it could lead to very weird edge cases. I think we should just raise an error.

File: nilearn/tests/test_masking.py [hunk 3]
```diff
@@ -351,3 +355,5 @@ def test_warning_shape(random_state=42, shape=(3, 5, 7, 11)):
     X = rng.randn(n_samples, n_features)
     # 2D X (should be ok)
     unmask(X, mask_img)
+
```
Comment (lesteve) [inline nilearn/tests/test_masking.py]: Could you remove these two additional lines?

File: nilearn/tests/test_masking.py [hunk 4]
```diff
@@ -331,7 +335,7 @@ def test_compute_multi_epi_mask():
     assert_array_equal(mask_ab, mask_ab_.get_data())
 
 
-def test_warning_shape(random_state=42, shape=(3, 5, 7, 11)):
+def test_error_shape(random_state=42, shape=(3, 5, 7, 11)):
```
Comment (lesteve) [inline nilearn/tests/test_masking.py]: Any particular reason why you rename this function?

File: nilearn/masking.py [hunk 5]
```diff
@@ -665,7 +665,15 @@ def _unmask_nd(X, mask, order="C"):
     if mask.dtype != np.bool:
         raise ValueError("mask must be a boolean array")
     if X.ndim != 2:
-        raise ValueError("X must be a 2-dimensional array")
+        raise TypeError("X must be a 2-dimensional array")
+    n_features = mask.sum()
+    if X.shape[-1] != n_features:
```
Comment (eickenberg) [inline nilearn/masking.py]: Nitpick: Can we have `X.shape[1]` here? (By the previous checks, `X` must be 2D, and it is better to be explicit)

File: nilearn/masking.py (hunk 2 above)
Comment (eickenberg) [inline nilearn/masking.py]: I would prefer `if X.shape[0] == n_features and X.shape[1] != n_features` in order not to have inexplicable transposes one day somebody uses a lot of data or a very small mask.

Comment (eickenberg) [inline nilearn/masking.py]: Or what Alex says, didn't see that.

File: nilearn/tests/test_masking.py (hunk 4 above)
Comment (eickenberg) [inline nilearn/tests/test_masking.py]: because now it raises an error?

**Comment by eickenberg** [comment]
While you are at it, could you look at the docstring of `_unmask_3D`, I think the shape of the input is wrong.

Also, I would propose to rename `_unmask_nd` to `_unmask_4d`, because it doesn't do any n-d at all and may confuse people trying to understand the code.

File: nilearn/tests/test_masking.py (hunk 4 above)
Comment (bcipolli) [inline nilearn/tests/test_masking.py]: This tests errors, not warnings.

**Comment by bcipolli** [comment]
Thanks for the comments!  Should all be set.

**Comment by eickenberg** [comment]
Could you change l628 from `(n_samples,)` to `(n_features,)`? (Or am I wrong about this?)

**Comment by bcipolli** [comment]
@eickenberg good eye!  Just pushed that change now as well.

File: nilearn/tests/test_masking.py (hunk 4 above)
Comment (lesteve) [inline nilearn/tests/test_masking.py]: Right I didn't properly expand the diff and only saw the "2D X (should be ok) test", sorry for that

**Comment by lesteve** [comment]
You have one test failing, probably the one with the auto transpose logic that was removed.

File: nilearn/masking.py (hunk 1 above)
Comment (GaelVaroquaux) [inline nilearn/masking.py]: +1

File: nilearn/masking.py [hunk 6]
```diff
@@ -665,7 +665,10 @@ def _unmask_nd(X, mask, order="C"):
     if mask.dtype != np.bool:
         raise ValueError("mask must be a boolean array")
```
Comment (banilo) [inline nilearn/masking.py]: rather TypeError?

**Comment by bcipolli** [comment]
@lesteve Fixed that.
@banilo Changed that and added tests (there were none).
Also, while creating new tests, noticed that the 3D and 4D errors for wrong shape were different.  Added the similar shape check to the 3D code, and updated the tests.

**Comment by AlexandreAbraham** [comment]
Thanks !
